File: 0621/Exercise03.js

class Citizen {
  name;
  age;
  
  constructor(name, age){
    this.name = name;
    this.age = age;
  }
}

let kim = new Citizen('Kim', 22);
let lee = new Citizen('Lee', 25);
let park = new Citizen('Park', 18);
let jin = new Citizen('Jin', 16);

let citizens = [kim, lee, park, jin];

// let citizens_copy = citizens.entries(); // interate
let citizens_copy = [...citizens];         //array

console.log('전체 시민들: ');
citizens.forEach(citizen => console.log(`${JSON.stringify(citizen)} `));

console.log('투표권이 있는 시민들: ');
citizens.filter(adult => adult.age > 18)
          .forEach(citizen => console.log(`${JSON.stringify(citizen)} `));

console.log('내년에 투표권이 생기는 시민들: ');

citizens.filter(citizen => citizen.age === 18)
  .forEach(citizen => console.log(`${JSON.stringify(citizen)} `));
  
let numList = [ 2, 53, 1, 0, 4, 10 ];
console.log('flatted list: ', numList);

numList = numList.reverse();
console.log('reversed list: ', numList);

numList = numList.sort();

console.log('sorted list: ', numList);

/**
 * 
 * @param {*} a 
 * @param {*} b 
 * a is less than b : -1
 * a is greater than b : 1
 * a must be equal to b : 0
 */

 
let sortOption = (a,b) => {
  let val = a < b ? -1 : a < b ? 1 : 0;
  return val;
}

numList = numList.sort(sortOption);

console.log('sorted list2: ', numList);

File: 0621/Exercise03.cpp
/***
 * 연결 리스트에서 remove_if() 함수를 이용한 조건부 원소 삭제
 * 
*/

#include <string>
#include <iostream>
#include <forward_list>

struct citizen
{
  std::string name;
  int age;
};

std::ostream &operator<<(std::ostream &os, const citizen &c)
{
  return (os << "[" << c.name << ", " << c.age << "]");
}

int main() {
  std::forward_list<citizen> citizens = {
    {"Kim", 22}, {"Lee", 25}, {"Park", 18}, {"Jin", 16}
  };
  
  auto citizens_copy = citizens;      //깊은 복사
  
  std::cout << "전체 시민들: ";
  for (const auto &c : citizens)
    std::cout << c << " ";
  std:: cout << std::endl;
  
  citizens.remove_if([](const citizen &c) {
    return (c.age < 19);
  });
  
  std::cout << "투표권이 있는 시민들: ";
  for (const auto &c : citizens)
    std::cout << c << " ";
  std::cout << std::endl;
  
  citizens_copy.remove_if([](const citizen &c) {
    return (c.age != 18);
  });
  
  std::cout << "내년에 투표권이 생기는 시민들: ";
  for (const auto &c : citizens_copy)
    std::cout << c << " ";
  std::cout << std::endl;
  
  std::forward_list<int> list1 = { 2, 53, 1, 0, 4, 10 };
  
  list1 = { 0,1,0,01,1,10,5,5,10};
  
  list1.reverse();
  
  list1.sort();
  list1.unique();
  
  list1.unique([](int a, int b) { return (b - a) < 2; });
  
}
